### PR TITLE
Feature/126 tales and villains deckbuilding restrictions

### PR DESCRIPTION
**Pull request by Niebvelungen**
Implementation for  #126 

Requirements:
- migrate DB
-------------------------------------------------------------------------------------------------------------------------------------------
### Needed measures to get functionallity: [Tales] - [Villain] restriction
- create tag "Tales" + "Villains" in Admin site
- tag tales and villain cards with corresponding tag (can be done via sql insert script)
- create "restriction" in Admin site 
image
- create new restriction_action for the restriciton with technical_name "conflicting_tag"

image

(Technical_name is used to define what action should happen in js. Current supported actions are: conflicting_tag,  singelton (does not check sideboard or stones for dupplicates), arcana_singelton)

**Done**
Now if there are Tales and Villains in any zone except the Sideboard they will be marked and a warning will be displayed.

-------------------------------------------------------------------------------------------------------------------------------------------

### Needed measures to get functionallity: Singelton restriction
- create tag "Singelton" + "Singelton_Override" in Admin site
- tag singelton Rulers and cards that overrule singelton (Basic Stones, Crocoshark) with corresponding tag (can be done via sql insert script)
- create "restriction" in Admin site 
image
- create new restriction_action for restriction with technical_name singelton

image

**Done**
Now if the singelton tagged card is in the ruler area it will apply the singelton rule to every card that does not have the override tag or is in excluded zones (Side Board, Magic Stones)

-------------------------------------------------------------------------------------------------------------------------------------------

### Needed measures to get functionallity: Arcana Singelton restriction
- create tag "Arcana Singelton" + "Singelton_Override" (can be reused if already available since it works the same) in Admin site
- tag singelton Rulers and cards that overrule singelton (Basic Stones, Crocoshark) with corresponding tag (can be done via sql insert script)
- create "restriction" in Admin site 
image
- create new restriction_action for restriction with technical_name arcana_singelton

image

**Done**
Now every card in every zone that does not have Singelton_Override and violates the singelton rule will be marked and a warning displayed

### CODE REVIEW

File: cardDatabase/static/js/view_decklist_restricted.js [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+                restrictionObject.applyAction();
+                registratedRestrictions.push(restriction.action);
+            }
+            else
+                console.error(`Attempted duplicate registration of restriction ${restriction.action}`);
+        });
+    }
+}
+
+class RestricitonFactory {
```
Comment (Kossetsoup) [inline cardDatabase/static/js/view_decklist_restricted.js:23]: Restriction misspelled here as Restriciton, and many other places!

File: cardDatabase/static/js/view_decklist_restricted.js [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    restrictedToZones = ['Ruler', 'Ruler Area', 'Arcana Ruler'];
+    ignoredZones = ['Side', 'Side Deck', 'Side Board', 'Side Board Deck', 'Magic', 'Magic Stones', 'Magic Stone Deck'];
+
+    getRestrictionForAction(technicalName, cardContainer, cardData, tagToCheck, restrictedTag, warninigText, warningTextOutputElement) {
+        switch (technicalName) {
+            case 'conflicting_tag':
+                return new ConflictingTagRestriciton(cardContainer, cardData, tagToCheck, restrictedTag, warninigText, warningTextOutputElement, [], this.ignoredZones);
+            case 'singelton':
```
Comment (Kossetsoup) [inline cardDatabase/static/js/view_decklist_restricted.js:31]: singleton not singelton

File: cardDatabase/static/js/view_decklist_restricted.js [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                            <img class="banned-icon" src="${$('#banned_icon').attr('src')}">
+                            <div class="ban-text">${warninigText}</div>
+                        </div>`);
+    }
+
+    distinctFilter = function (value, index, array) {
+        return array.indexOf(value) === index;
+    }
+
+    highLightRestrictedCards = function (affectedCards, tooltip) {
```
Comment (Kossetsoup) [inline cardDatabase/static/js/view_decklist_restricted.js:76]: highlight is one word, L shouldn't be capitalised

File: cardDatabase/migrations/0046_auto_20230802_0832.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        ),
+        migrations.CreateModel(
+            name='Restriction',
+            fields=[
+                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('text', models.TextField()),
+                ('last_modified', models.DateTimeField(auto_now_add=True)),
+                ('action', models.ForeignKey(null=True, on_delete=django.db.models.deletion.CASCADE, to='cardDatabase.restrictionaction')),
+                ('added_by', models.ForeignKey(null=True, on_delete=django.db.models.deletion.SET_NULL, to='cardDatabase.profile')),
+                ('restrictioned_tag', models.ForeignKey(null=True, on_delete=django.db.models.deletion.CASCADE, to='cardDatabase.tag')),
```
Comment (Kossetsoup) [inline cardDatabase/migrations/0046_auto_20230802_0832.py:37]: restrictioned is not a word, we want "restricted"

**Comment by Niebvelungen** [comment]
@Kossetsoup fixed comments

**Comment by Niebvelungen** [comment]
It is now possible to create Restriction Exceptions for a Restriction.
The following types are supported: `full_exception` and `partial_exception`.

### Creating a Exception:

1. Find the designated cards where the exception is used for our example i will use Zeus, first Olympian and Zeus Enters the Game of Gods (Zeus is a singleton Deck that allows 4x Copies of Zeus Enters ...)

2. Create the Restriction Exception in the Admin Site 

- First Select the restriction for which the exception is being built
- Then select the card that applys the exception (in our case Zeus, first Olympian)
- Define the Zone the Card has to be in to apply said exception (multiple zones can be define separated by `;`)
- Choose or create a new `Exception Action` that takes place for the current exception

2.5 Creating a new Exception Action

- Give the exception a fitting name ("Exception Zeus Enters GoG" in our case)
- Select the action that the exception applies:
          - `full_exception` disables the restiction completly when the criteria of the assigned `RestrictionException` are met
          - `partial_exception` only affects cards that are selected in the next step (this would be the one we need for Zeus) 
- Select the cards that the exception is applied to (this is needed for `partial_exception`)

Here screenshot on how our exception will look like after configuring

image
image

Here the Screenshot of it in full action

image

**Review by Kossetsoup — APPROVED** [review]
Good work, I like the new exemption feature! My one criticism of it is that the exemptions are reliant on the names of the zones, which works fine if you keep the default zone names which many people will - but some people might like to rename them. 

I think for now it is fine but it may be worth another look in the future (a way to tag zones you've created as "Main" or what have you).

**Pull request merged**